src/Microsoft.OData.Core/Json/ODataJsonWriterUtils.cs

```diff
@@ -98,32 +99,52 @@ internal static void ODataValueToString(StringBuilder sb, ODataValue value)
             if (value == null || value is ODataNullValue)
             {
                 sb.Append("null");
+                return;
             }
 
-            ODataCollectionValue collectionValue = value as ODataCollectionValue;
-            if (collectionValue != null)
+            if (value is ODataCollectionValue collectionValue)
             {
                 ODataCollectionValueToString(sb, collectionValue);
+                return;
             }
 
-            ODataResourceValue resourceValue = value as ODataResourceValue;
-            if (resourceValue != null)
+            if (value is ODataResourceValue resourceValue)
             {
                 ODataResourceValueToString(sb, resourceValue);
+                return;
             }
 
-            ODataPrimitiveValue primitiveValue = value as ODataPrimitiveValue;
-            if (primitiveValue != null)
+            if (value is ODataPrimitiveValue primitiveValue)
             {
-                if (primitiveValue.FromODataValue() is string)
+                object valueAsObject = primitiveValue.FromODataValue();
+                string valueAsString;
+                if (ODataRawValueUtils.TryConvertPrimitiveToString(valueAsObject, out valueAsString))
                 {
-                    sb.Append(string.Concat("\"", JsonValueUtils.GetEscapedJsonString(value.FromODataValue()?.ToString()), "\""));
+                    if (valueAsObject is string)
+                    {
+                        valueAsString = JsonValueUtils.GetEscapedJsonString(valueAsString);
+                    }
+
+                    if (valueAsObject is string || valueAsObject is byte[] || valueAsObject is DateTimeOffset || valueAsObject is Guid || valueAsObject is TimeSpan | valueAsObject is Date || valueAsObject is TimeOfDay)
+                    {
+                        sb.Append('"').Append(valueAsString).Append('"');
+                    }
+                    else
+                    {
+                        sb.Append(valueAsString);
+                    }
                 }
-                else
-                {
-                    sb.Append(JsonValueUtils.GetEscapedJsonString(value.FromODataValue()?.ToString()));
+                else
+                {
+                    // In the unlikely event
+                    sb.Append('"').Append(JsonValueUtils.GetEscapedJsonString(ODataErrorStrings.ODataJsonWriter_UnsupportedValueType(value.GetType().FullName))).Append('"');
                 }
+
+                return;
             }
+
+            // Subclasses of ODataValue that are not supported in ODataInnerError.Properties dictionary
+            sb.Append('"').Append(JsonValueUtils.GetEscapedJsonString(ODataErrorStrings.ODataJsonWriter_UnsupportedValueType(value.GetType().FullName))).Append('"');

```

Comment:
What was the concern with throwing an exception in this case?

Reply:
Throwing an exception from `ToString()` is considered a very bad idea. The logic in this method is only executed when `ODataError.ToString()` and `ODataInnerError.ToJsonString()` are executed.